src/sync/mod.rs

```diff
@@ -0,0 +1,340 @@
+
+use crate::{
+    api::{AnchoringProposalState, PrivateApi},
+    blockchain::SignInput,
+    btc,
+    config::Config,
+};
+
+mod bitcoin_relay;
+
+/// Anchoring transaction with its index in the anchoring chain.
+pub type TransactionWithIndex = (btc::Transaction, u64);
+
+type KeyPool = Arc<HashMap<btc::PublicKey, btc::PrivateKey>>;
+
+/// Errors that occur when updating the anchoring chain.
+#[derive(Debug)]
+pub enum ChainUpdateError<C: Display> {
+    /// Error occurred in the private API client.
+    Client(C),
+    /// Insufficient funds to create an anchoring transaction proposal.
+    InsufficientFunds {
+        /// Total transaction fee.
+        total_fee: u64,
+        /// Available balance.
+        balance: u64,
+    },
+    /// Internal error.
+    Internal(failure::Error),
+}
+
+/// Signs the inputs of the anchoring transaction proposal by the corresponding
+/// Bitcoin private keys.
+#[derive(Debug)]
+pub struct AnchoringChainUpdateTask<T>
+where
+    T: PrivateApi + 'static,
+{
+    key_pool: KeyPool,
+    api_client: T,
+}
+
+impl<T> AnchoringChainUpdateTask<T>
+where
+    T: PrivateApi + 'static,
+    T::Error: Display,
+{
+    /// Creates a new anchoring chain updater instance.
+    pub fn new(
+        keys: impl IntoIterator<Item = (btc::PublicKey, btc::PrivateKey)>,
+        api_client: T,
+    ) -> Self {
+        Self {
+            key_pool: Arc::new(keys.into_iter().collect()),
+            api_client,
+        }
+    }
+
+    /// Performs one attempt to sign an anchoring proposal, if any.
+    pub fn process(&self) -> Result<(), ChainUpdateError<T::Error>> {
+        log::trace!("Perform an anchoring chain update");
+
+        match self
+            .api_client
+            .anchoring_proposal()
+            .map_err(ChainUpdateError::Client)?
+        {
+            AnchoringProposalState::None => Ok(()),
+            AnchoringProposalState::Available {
+                transaction,
+                inputs,
+            } => {
+                let config = self.api_client.config().map_err(ChainUpdateError::Client)?;
+                self.handle_proposal(config, transaction, inputs)
+            }
+            AnchoringProposalState::InsufficientFunds { balance, total_fee } => {
+                Err(ChainUpdateError::InsufficientFunds { balance, total_fee })
+            }
+        }
+    }
+
+    fn handle_proposal(
+        &self,
+        config: Config,
+        proposal: btc::Transaction,
+        inputs: Vec<btc::Transaction>,
+    ) -> Result<(), ChainUpdateError<T::Error>> {
+        log::trace!("Got an anchoring proposal: {:?}", proposal);
+        // Find among the keys one from which we have a private part.
+        // TODO What we have to do if we find more than one key? [ECR-3222]
+        let keypair = if let Some(keypair) =
+            self.find_private_key(config.anchoring_keys.iter().map(|x| x.bitcoin_key))
+        {
+            keypair
+        } else {
+            return Ok(());
```

Comment:
Why is it OK if we don't find a key?

Reply:
For example we are waiting until the anchoring configuration changed to add us as an anchoring node.

Reply:
Maybe add a comment then?